File: leetcode/0454-4sum-II/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 276 ms (90.38%)
     * MEMORY: 27.5 MB (15.64%)
    */
    int fourSumCount(vector<int>& nums1, vector<int>& nums2, vector<int>& nums3, vector<int>& nums4) {
        unordered_map <int, int> sum1, sum2;
        int total = 0;
        for (int i=0; i < nums1.size(); i++ ) {
            for (int j=0; j < nums1.size(); j++ ) {
                int tmp1 = nums1[i] + nums2[j];
                int tmp2 = nums3[i] + nums4[j];
                ++sum1[tmp1];
                ++sum2[tmp2];
            }
        }
        for (auto i : sum1) {
            if (sum2.find(-i.first) != sum2.end()) {
                total += i.second * sum2[-i.first];
            }
        }
        return total;
    }
};

File: leetcode/0454-4sum-II/solution.py
class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        """
        RUNTIME: 878 ms (76.24%)
        MEMORY: 14.2 MB (71.89%)
        """
        sum1, sum2 = {}, {}
        total = 0
        for i in range(len(nums1)):
            for j in range(len(nums1)):
                tmp1 = nums1[i] + nums2[j]
                tmp2 = nums3[i] + nums4[j]
                if tmp1 not in sum1:
                    sum1[tmp1] = 1
                else:
                    sum1[tmp1] += 1
                if tmp2 not in sum2:
                    sum2[tmp2] = 1
                else:
                    sum2[tmp2] += 1

        if (len(sum1) > len(sum2)):
            sum1, sum2 = sum2, sum1

        for i in sum1:
            if -i in sum2:
                total += sum1[i] * sum2[-i]
        return total
